worlds/shapez/locations.py

```diff
@@ -0,0 +1,578 @@
+    for shape in ["Circle", "Square", "Star", "Windmill"]:
+        for color in color_names:
+            color_region = color_to_needed_building([color])
+            shapesanity_simple[f"Half {color} {shape}"] = f"Shapesanity Half {color_region}"
+            shapesanity_simple[f"{color} {shape} Piece"] = f"Shapesanity Piece {color_region}"
+            shapesanity_simple[f"Cut Out {color} {shape}"] = f"Shapesanity Stitched {color_region}"
+            shapesanity_simple[f"Cornered {color} {shape}"] = f"Shapesanity Stitched {color_region}"
+    # one color && 4 shapes (including empty)
+    for first_color in ["r", "g", "b", "y", "p", "c"]:
+        for second_color in ["g", "b", "y", "p", "c", "w"]:
+            if not first_color == second_color:
+                for third_color in ["b", "y", "p", "c", "w", "u"]:
+                    if third_color not in [first_color, second_color]:
+                        for fourth_color in ["y", "p", "c", "w", "u"]:
+                            if fourth_color not in [first_color, second_color, third_color]:
+                                colors = [first_color, second_color, third_color, fourth_color]
+                                color_region = color_to_needed_building(colors)
+                                color_code = ''.join(sorted(colors))
+                                shapesanity_1_4[f"{color_code} Circle"] = f"Shapesanity Colorful Full {color_region}"
+                                shapesanity_1_4[f"{color_code} Square"] = f"Shapesanity Colorful Full {color_region}"
+                                shapesanity_1_4[f"{color_code} Star"] = f"Shapesanity Colorful Full {color_region}"
+                                shapesanity_1_4[f"{color_code} Windmill"] \
+                                    = f"Shapesanity Colorful East Windmill {color_region}"
+                        fourth_color = "-"
+                        colors = [first_color, second_color, third_color, fourth_color]
+                        for shape in ["Circle", "Square", "Windmill", "Star"]:
+                            shapesanity_1_4[f"{''.join(sorted(colors))} {shape}"] \
+                                = f"Shapesanity Stitched {color_to_needed_building(colors)}"
+    # one shape && 4 colors (including empty)
+    for color in color_names:
+        for first_shape in ["C", "R"]:
+            for second_shape in ["R", "W"]:
+                if not first_shape == second_shape:
+                    for third_shape in ["W", "S"]:
+                        if not third_shape == second_shape:
+                            for fourth_shape in ["S", "-"]:
+                                if not fourth_shape == third_shape:
+                                    shapes = [first_shape, second_shape, third_shape, fourth_shape]
+                                    shapesanity_1_4[f"{color} {''.join(sorted(shapes))}"] \
+                                        = f"Shapesanity Stitched {color_to_needed_building([color])}"
```

Comment:
In fact, while there is a performance benefit here, which is that checking `if not first_shape == second_shape` *before* even going into the for loop for the third shape makes it "early exit" a lot better, I wonder how much that is actually worth over the readability of something like:

```
for first_shape, second_shape, third_shape, fourth_shape in itertools.product(["C", "R"], ["R", "W"], ["W", "S"], ["S", "-"]):
    if len(set(first_shape, second_shape, third_shape, fourth_shape)) < 4:
        continue
    ... rest of the code ...
```

Or maybe you could set up your own helper function that does "itertools.product, but without duplicates" efficiently

Reply:
In fact, is what you're doing here not just a reimplementation of `itertools.combinations`?

Reply:
Although I imagine some of the more complicated chains lower down in this file can't be easily done using something like that.

Either way, I'd encourage some use of some of these functions if possible

Reply:
Didn't know about itertools, it's the first time for me coding more than a few lines in python (except for a simple message-to-keystroke discord bot), so I only know very few builtin tools.

I could find a use for both `itertools.combinations()` and `itertools.permutations()` and also implemented the other suggestions, but the time for calculating the shapesanity pool went from ~1.3 to ~1.5 seconds (at least on my PC).